pkg/server/cache.go

```diff
@@ -0,0 +1,141 @@
+	cacheserver "github.com/kcp-dev/kcp/pkg/cache/server"
+	cacheoptions "github.com/kcp-dev/kcp/pkg/cache/server/options"
+)
+
+// installCacheServer integrates the cache-server into the kcp-server.
+// The new server is available at /services/cache/... HTTP path.
+func (s *Server) installCacheServer(ctx context.Context) error {
+	cacheServerOptions := cacheoptions.NewOptions(s.Options.Extra.RootDirectory)
+	cacheServerEtcdOptions := *s.Options.GenericControlPlane.Etcd
+	cacheServerOptions.Etcd = &cacheServerEtcdOptions
+	cacheServerOptions.SecureServing = s.Options.GenericControlPlane.SecureServing
+	completedCacheServerOptions, err := cacheServerOptions.Complete()
+	if err != nil {
+		return err
+	}
+
+	restConfig := rest.CopyConfig(s.GenericConfig.LoopbackClientConfig)
+	restConfig = withCacheServiceRoundTripper(restConfig)
+	newCacheServerConfig, err := cacheserver.NewConfig(completedCacheServerOptions, restConfig)
+	if err != nil {
+		return err
+	}
+	completedCacheServerConfig, err := newCacheServerConfig.Complete()
+	if err != nil {
+		return err
+	}
+	newCacheServer, err := cacheserver.NewServer(completedCacheServerConfig)
+	if err != nil {
+		return err
+	}
+	completedCacheServer, err := newCacheServer.Complete(ctx)
+	if err != nil {
+		return err
+	}
+
+	preparedCacheServer := completedCacheServer.PrepareRun()
+	if err := s.AddPostStartHook("kcp-start-cache-server", func(ctx genericapiserver.PostStartHookContext) error {
+		preparedCacheServer.RunPostStartHooks(ctx.StopCh)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	s.preHandlerChainMux.Handle(virtualcommandoptions.DefaultRootPathPrefix+"/cache/", withServiceScope(preparedCacheServer.Handler))
+	return nil
+}
+
+// withServiceScope an HTTP filter that trims "/services/cache" prefix from the URL.
+//
+// for example: /services/cache/shards/amber/clusters/*/apis/apis.kcp.dev/v1alpha1/apiexports
+// is truncated to /shards/amber/clusters/*/apis/apis.kcp.dev/v1alpha1/apiexports
+func withServiceScope(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if path := req.URL.Path; strings.HasPrefix(path, "/services/cache") {
+			path = strings.TrimPrefix(path, "/services/cache")
+			req.URL.Path = path
+			newURL, err := url.Parse(req.URL.String())
+			if err != nil {
+				responsewriters.ErrorNegotiated(
+					apierrors.NewInternalError(fmt.Errorf("unable to resolve %s, err %w", req.URL.Path, err)),
+					errorCodecs, schema.GroupVersion{},
+					w, req)
+				return
+			}
+			req.URL = newURL
+		}
+		handler.ServeHTTP(w, req)
+	})
+}
+
+// withCacheServiceRoundTripper wraps an existing config's with cacheServiceRoundTripper.
+func withCacheServiceRoundTripper(cfg *rest.Config) *rest.Config {
+	cfg.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+		return newCacheServiceRoundTripper(rt)
+	})
+	return cfg
+}
+
+// cacheServiceRoundTripper is a http.RoundTripper that appends "/services/cache" prefix to a request.
```

Comment:
I think that the cache-server should add the `/services/cache` prefix. It would always give us the same path, whether we run as a standalone server or part of kcp. WDYT?